File: CPP/ByTier/B/2168_B_Locate.cpp
// Problem: CF 2168 B - Locate
// https://codeforces.com/contest/2168/problem/B

/*
 * Problem: Locate
 * Algorithm: Binary search with pre-processing
 * 
 * Approach:
 * - First run (Player A): Determine x based on relative positions of n and 1
 * - Second run (Player B): Use binary search to find position of n based on x value
 * 
 * Time Complexity: O(n log n) for first run (preprocessing) + O(log n) for second run
 * Space Complexity: O(1) - only using a few variables
 *
 * Strategy:
 * - Player A: Check if n occurs before or after position of 1 in permutation
 *   - If n is before 1, send 0
 *   - If n is after 1, send 1
 * - Player B: 
 *   - If x=0: Binary search from left to find position using query range [1, med]
 *   - If x=1: Binary search from right to find position using query range [med, n]
 */

#include <iostream>
#include <string>

using namespace std;
using ll = long long;
const int mod = 998244353;
const int block_size = 7;
string run_id;
string s;
int query(int st, int dr) {
  cout << "? " << st << " " << dr << endl;
  int r;
  cin >> r;
  return r;
}
int target;
int bin_src_dr(int st, int dr) {
  int med, poz = dr;
  while (st <= dr) {
    int med = (st + dr) >> 1;
    if (query(1, med) == target) {  // Check if max-min in range [1, med] equals target
      poz = med;
      dr = med - 1;
    } else
      st = med + 1;
  }
  return poz;
}
int bin_src_st(int st, int dr) {
  int med, poz = 1;
  while (st <= dr) {
    int med = (st + dr) >> 1;
    if (query(med, target + 1) == target) {  // Check if max-min in range [med, target+1] equals target
      poz = med;
      st = med + 1;
    } else
      dr = med - 1;
  }
  return poz;
}
int main() {
  int n, i, t, x, y, k;
  cin >> run_id;
  if (run_id == "first") {
    cin >> t;
    while (t--) {
      cin >> n;
      int p, p1;
      for (i = 1; i <= n; i++) {
        cin >> x;
        if (x == n)
          p = i;   // Store position of n
        if (x == 1)
          p1 = i;  // Store position of 1
      }
      if (p <= p1)
        cout << "0\n";  // If n occurs before or at same position as 1, send 0
      else
        cout << "1\n";  // If n occurs after 1, send 1
    }
  } else {
    cin >> t;
    int f;
    while (t--) {
      cin >> n >> f;
      target = n - 1;  // target represents difference between max and min in range
      if (!f) {
        int rez = bin_src_st(1, n);  // Binary search from left
        cout << "! " << rez << endl;
      } else {
        int rez = bin_src_dr(1, n);  // Binary search from right
        cout << "! " << rez << endl;
      }
    }
  }
  return 0;
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByRound/2168/2168_B_Locate.py
# Problem: CF 2168 B - Locate
# https://codeforces.com/contest/2168/problem/B

"""
Problem: Locate
Algorithm/Technique: Binary search, interactive problem solving
Time Complexity: O(n log n) for first run, O(log n) for second run
Space Complexity: O(n) for storing the permutation

This is a run-twice (communication) interactive problem where:
- Player A (first run) receives the full permutation and sends a binary value x to jury
- Player B (second run) receives only n and x, then uses at most 30 queries to find
  the position of integer n in the permutation

Strategy:
- First run (Player A): 
  - Determine x = 0 if the position of 1 is less than position of n
  - Otherwise, x = 1
- Second run (Player B):
  - Use binary search to locate n's position
  - When x = 0: search for the leftmost position where query(0, pos) == n-1
  - When x = 1: search for the rightmost position where query(pos, n) != n-1

The difference between max and min in a subarray gives us information about the range
in that subarray. When we know n's position, we can use this to deduce its location
by checking subarrays containing or not containing n.
"""

import os
import sys
from io import BytesIO, IOBase


def discrete_binary_search(func, lo, hi):
    """Locate the first value x s.t. func(x) = True within [lo, hi]"""
    while lo < hi:
        mi = lo + (hi - lo) // 2
        if func(mi):
            hi = mi
        else:
            lo = mi + 1

    return lo


def main():
    def query(l, r):
        print("?", l + 1, r, flush=True)
        resp = int(input())
        assert resp != -1
        return resp

    run = input()

    if run == "first":
        # First run - Player A
        t = int(input())
        for _ in range(t):
            n = int(input())
            # Convert to 0-based indexing for easier handling
            p = [int(pi) - 1 for pi in input().split()]
            # Send x = 0 if position of 1 (0-indexed) is less than position of n (n-1-indexed)
            # Otherwise send x = 1
            print(0 if p.index(0) < p.index(n - 1) else 1)
    else:
        # Second run - Player B
        t = int(input())
        for _ in range(t):
            n, x = map(int, input().split())

            if x == 0:
                # If x = 0, then position of 1 is less than position of n
                # We want to find the leftmost position such that range [0, pos] contains n
                # and has range (n-1). So we are looking for first such position.
                print(
                    "!",
                    discrete_binary_search(lambda x: query(0, x) == n - 1, 2, n),
                    flush=True,
                )
            else:
                # If x = 1, then position of 1 is greater than or equal to position of n
                # We look for a position such that range [pos, n] does not contain n
                # This means the range is not n-1, so n must be to the left of pos
                print(
                    "!",
                    discrete_binary_search(lambda x: query(x, n) != n - 1, 1, n - 1),
                    flush=True,
                )


BUFSIZE = 8192


class FastIO(IOBase):
    newlines = 0

    def __init__(self, file):
        self._file = file
        self._fd = file.fileno()
        self.buffer = BytesIO()
        self.writable = "x" in file.mode or "r" not in file.mode
        self.write = self.buffer.write if self.writable else None

    def read(self):
        while True:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            if not b:
                break
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines = 0
        return self.buffer.read()

    def readline(self):
        while self.newlines == 0:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            self.newlines = b.count(b"\n") + (not b)
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines -= 1
        return self.buffer.readline()

    def flush(self):
        if self.writable:
            os.write(self._fd, self.buffer.getvalue())
            self.buffer.truncate(0), self.buffer.seek(0)


class IOWrapper(IOBase):
    def __init__(self, file):
        self.buffer = FastIO(file)
        self.flush = self.buffer.flush
        self.writable = self.buffer.writable
        self.write = lambda s: self.buffer.write(s.encode("ascii"))
        self.read = lambda: self.buffer.read().decode("ascii")
        self.readline = lambda: self.buffer.readline().decode("ascii")


sys.stdin, sys.stdout = IOWrapper(sys.stdin), IOWrapper(sys.stdout)
input = lambda: sys.stdin.readline().rstrip("\r\n")


if __name__ == "__main__":
    main()


# https://github.com/VaHiX/CodeForces/